File: src/JAVA/DetectaCaras.java
import org.bytedeco.javacv.*;
import org.bytedeco.opencv.opencv_core.*;
import org.bytedeco.opencv.opencv_objdetect.*;

public class DetectaCaras {
    public static void main(String[] args) throws Exception {
        // Inicializar el capturador de video
        OpenCVFrameGrabber grabber = new OpenCVFrameGrabber(0);
        grabber.start();

        // Cargar el clasificador de detección de rostros
        CascadeClassifier classifier = new CascadeClassifier();
        classifier.load("haarcascade_frontalface_default.xml");

        // Crear una ventana para mostrar el video en vivo
        CanvasFrame frame = new CanvasFrame("Face Detection");
        frame.setDefaultCloseOperation(javax.swing.JFrame.EXIT_ON_CLOSE);

        // Bucle principal
        while (frame.isVisible()) {
            // Capturar el siguiente fotograma del video
            Frame videoFrame = grabber.grab();

            // Convertir el fotograma en una matriz de píxeles
            opencv_core.Mat mat = new opencv_core.Mat(videoFrame);
            opencv_core.Mat grayMat = new opencv_core.Mat();

            // Convertir la matriz a escala de grises
            opencv_imgproc.cvtColor(mat, grayMat, opencv_imgproc.CV_BGR2GRAY);

            // Detectar rostros en la matriz de escala de grises
            RectVector faces = new RectVector();
            classifier.detectMultiScale(grayMat, faces);

            // Dibujar rectángulos alrededor de los rostros detectados
            for (int i = 0; i < faces.size(); i++) {
                Rect face = faces.get(i);
                opencv_imgproc.rectangle(mat, face, new Scalar(0, 255, 0, 1));
            }

            // Mostrar el fotograma con los rectángulos de los rostros
            frame.showImage(videoFrame);
        }

        // Detener el capturador de video y cerrar la ventana
        grabber.stop();
        frame.dispose();
    }
}

File: src/JAVA/DetectaCaras.py
import cv2
import face_recognition

# Cargar la imagen de referencia
reference_image = face_recognition.load_image_file("referencia.jpg")
reference_encoding = face_recognition.face_encodings(reference_image)[0]

# Inicializar la cámara
video_capture = cv2.VideoCapture(0)

while True:
    # Capturar un fotograma de la cámara
    ret, frame = video_capture.read()

    # Convertir el fotograma a RGB (necesario para face_recognition)
    rgb_frame = frame[:, :, ::-1]

    # Detectar rostros en el fotograma
    face_locations = face_recognition.face_locations(rgb_frame)
    face_encodings = face_recognition.face_encodings(rgb_frame, face_locations)

    # Comparar cada rostro detectado con la imagen de referencia
    for face_encoding in face_encodings:
        matches = face_recognition.compare_faces([reference_encoding], face_encoding)

        if True in matches:
            # Si hay una coincidencia, dibujar un cuadro alrededor del rostro
            top, right, bottom, left = face_locations[matches.index(True)]
            cv2.rectangle(frame, (left, top), (right, bottom), (0, 255, 0), 2)

    # Mostrar el fotograma resultante
    cv2.imshow('Video', frame)

    # Salir del bucle si se presiona la tecla 'q'
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Liberar los recursos
video_capture.release()
cv2.destroyAllWindows()
